tokenizer.go

```diff
@@ -67,6 +88,155 @@ func FromFile(path string) (*Tokenizer, error) {
 	return &Tokenizer{tokenizer: tokenizer}, nil
 }
 
+// LoadTokenizerFromHuggingFace downloads necessary files and initializes the tokenizer.
+// Parameters:
+//   - modelID: The Hugging Face model identifier (e.g., "bert-base-uncased").
+//   - destination: Optional. If provided and not nil, files will be downloaded to this folder.
+//     If nil, a temporary directory will be used.
+//   - authToken: Optional. If provided and not nil, it will be used to authenticate requests.
+func LoadTokenizerFromHuggingFace(modelID string, destination, authToken *string) (*Tokenizer, error) {
+	if strings.TrimSpace(modelID) == "" {
+		return nil, fmt.Errorf("modelID cannot be empty")
+	}
+
+	// Construct the model URL
+	modelURL := fmt.Sprintf("%s/%s/resolve/main", baseURL, modelID)
+
+	// Determine the download directory
+	var downloadDir string
+	if destination != nil && *destination != "" {
+		downloadDir = *destination
+		// Create the destination directory if it doesn't exist
+		err := os.MkdirAll(downloadDir, os.ModePerm)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create destination directory %s: %w", downloadDir, err)
+		}
+	} else {
+		// Create a temporary directory
+		tmpDir, err := os.MkdirTemp("", "huggingface-tokenizer-*")
+		if err != nil {
+			return nil, fmt.Errorf("error creating temporary directory: %w", err)
+		}
+		downloadDir = tmpDir
+	}
+
+	var wg sync.WaitGroup
+	errChan := make(chan error, len(tokenizerFiles))
+
+	// Mutex for synchronized logging
+	var logMutex sync.Mutex
+
+	// Download each tokenizer file concurrently
+	for filename, isMandatory := range tokenizerFiles {
+		wg.Add(1)
+		go func(fn string, mandatory bool) {
+			defer wg.Done()
+			fileURL := fmt.Sprintf("%s/%s", modelURL, fn)
+			destPath := filepath.Join(downloadDir, fn)
+			err := downloadFile(fileURL, destPath, authToken)
+			if err != nil {
+				if mandatory {
+					// If the file is mandatory, report an error
+					errChan <- fmt.Errorf("failed to download mandatory file %s: %w", fn, err)
+				} else {
+					// Optional files: log warning and continue
+					logMutex.Lock()
+					fmt.Printf("Warning: failed to download optional file %s: %v\n", fn, err)
+					logMutex.Unlock()
+				}
+			}
+		}(filename, isMandatory)
+	}
+
+	// Wait for all downloads to complete
+	wg.Wait()
+	close(errChan)
+
+	// Check for errors during downloads
+	for downloadErr := range errChan {
+		if downloadErr != nil {
```

Comment:
looks like all errors sent to the channel are non-nil so check is redundant. Just looking out for minimal indentation